### PR TITLE
Initial implementation of a parallel XML write option

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jptree.cpp [hunk 1]
```diff
@@ -6104,182 +6109,255 @@ inline bool isSanitizedAndHidden(const char *val, byte flags, bool attribute)
...
-    writeStringToStream(out, name);
-    Owned<IAttributeIterator> it = tree->getAttributes(true);
+public:
 
-    if (it->first())
+    IIOStream &out;
+    unsigned flags;
+    const IPropertyNodeVisitor *splitVisitor;
+    std::vector<const IPropertyTree *> stack;
+    Owned<CCompletionTask> completeTask;
```
Comment (ghalliday) [inline system/jlib/jptree.cpp:6120]: Should something wait for this complete?

File: system/jlib/jptree.cpp [hunk 2]
```diff
@@ -6104,182 +6109,255 @@ inline bool isSanitizedAndHidden(const char *val, byte flags, bool attribute)
...
+    void _toXML(const IPropertyTree *tree, unsigned indent, const IPropertyNodeVisitor *visitor, bool firstLevel)
     {
-        unsigned attributeindent = indent+2+(size32_t)strlen(name);
-        bool first = true;
+        const char *name = tree->queryName();
+        if (!name) name = "__unnamed__";
 
-        do
+        StringBuffer filename;
+        IPropertyNodeVisitor::VisitorAction action = IPropertyNodeVisitor::NoSplit;
```
Comment (ghalliday) [inline system/jlib/jptree.cpp:6142]: Not used outside of the if, so may as well move inside.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@richardkchapman A few comments/questions.

**Comment by ghalliday** [comment]
@jake best compared with whitespace ignored.

**Comment by richardkchapman** [comment]
Probably not the right approach

**Pull request closed without merging**